### PR TITLE
buildGoModule: warn if `-trimpath` or `-mod=` is used, remove explicit usage

### PR DESCRIPTION
**Pull request by katexochen**
## Description of changes

Follow up to https://github.com/NixOS/nixpkgs/pull/276685

Adding warnings if `-trimpath` or `-mod=` is explicitly set, as these flags are controlled by the builder.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by malt3 — APPROVED** [review]

File: pkgs/build-support/go/module.nix [hunk 1]
```diff
@@ -153,7 +155,7 @@ let
 
     inherit (go) GOOS GOARCH;
 
-    GOFLAGS = lib.optionals (!proxyVendor) [ "-mod=vendor" ] ++ lib.optionals (!allowGoReference) [ "-trimpath" ];
+    GOFLAGS = GOFLAGS ++ lib.optionals (!proxyVendor) [ "-mod=vendor" ] ++ lib.optionals (!allowGoReference) [ "-trimpath" ];
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/module.nix:158]:
```suggestion
    GOFLAGS = GOFLAGS
      ++ lib.optional (!proxyVendor) "-mod=vendor"
      ++ lib.optional (!allowGoReference) "-trimpath";
```
we can make this a little bit nicer

File: pkgs/build-support/go/package.nix [hunk 2]
```diff
@@ -89,7 +91,7 @@ let
 
     GO111MODULE = "off";
     GOTOOLCHAIN = "local";
-    GOFLAGS = lib.optionals (!allowGoReference) [ "-trimpath" ];
+    GOFLAGS = GOFLAGS ++ lib.optionals (!allowGoReference) [ "-trimpath" ];
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/package.nix:94]:
```suggestion
    GOFLAGS = GOFLAGS ++ lib.optional (!allowGoReference) "-trimpath";
```

File: pkgs/build-support/go/module.nix (hunk 1 above)
Comment (katexochen) [reply pkgs/build-support/go/module.nix]: Great, thanks! :)

**Review by malt3 — APPROVED** [review]
The change in uplosi is definitely fine. The overall change also looks good to me.

**Comment by katexochen** [comment]
@ofborg eval

**Review by kalbasit — APPROVED** [review]

**Comment by katexochen** [comment]
Rebased to resolve merge conflict. Also removed newly added `-trimpath` from `gcs` and `jq-lsp`.

**Review by jnsgruk — APPROVED** [review]
Looks good to me! thanks for the interations!

**Comment by katexochen** [comment]
@kalbasit @SuperSandro2000 @jnsgruk Could someone merge this?

**Comment by zowoq** [comment]
This can go to master instead of staging.

File: pkgs/build-support/go/module.nix (hunk 1 above)
Comment (zowoq) [reply pkgs/build-support/go/module.nix]: Don't agree that it is nicer. I'd prefer that this be reverted.

**Pull request merged**